Create ExecutionStats

abelsiqueira opened the pull request:
Breaking down #35

dpo commented on test/test_stats.jl:
```diff
@@ -0,0 +1,21 @@
+function test_stats()
+  stats = ExecutionStats(:first_order, obj=1.0, opt=1e-12, sol=ones(100),
```
is it `opt` or `dual_feas`?

dpo replied:
`sol` -> `solution`.

dpo commented on REQUIRE:
```diff
@@ -1,3 +1,3 @@
 julia 0.5
 Compat 0.7.15
-NLPModels 0.1.0
+NLPModels 0.2.0
```
this should be rebased

dpo commented on src/stats/stats.jl:
```diff
@@ -0,0 +1,144 @@
+      f = v
+    elseif k == :dual_feas || k == :opt
+      normg = v
+    elseif k == :elapsed_time || k == :time
```
I find it confusing that there would be several names for the same thing.

dpo commented on src/stats/stats.jl:
```diff
@@ -0,0 +1,144 @@
+    elseif k == :elapsed_time || k == :time
+      t = v
+    else
+      throw(UndefVarError(k))
```
Let's try to avoid `throw`ing because `catch`ing is slow.

dpo commented on src/stats/stats.jl:
```diff
@@ -0,0 +1,144 @@
+  @printf(io, "  solution: "); showvec(io, stats.solution); @printf(io, "\n")
+  @printf(io, "  iterations: "); show(io, stats.iter); @printf(io, "\n")
+  @printf(io, "  elapsed time: "); show(io, stats.elapsed_time); @printf(io, "\n")
+  if length(stats.solver_specific) > 0
```
Maybe this should be delegated to the solver-specific stats type?

dpo replied:
Maybe `ExecutionStats` should be `GenericExecutionStats <: AbstractExecutionStats`?

dpo commented on src/stats/stats.jl:
```diff
@@ -0,0 +1,144 @@
+  if length(x) == 0
+    @printf(io, "∅")
+  elseif length(x) <= 5
+    Base.show_delim_array(io, x, "", " ", "", false)
```
Great! I didn't know `show_delim_array`. How about `Base.show_delim_array(io, x, "[", " ", "]", false)` though?

abelsiqueira commented:
I rebased and made the changes, removing the various ways to pass a stat. Now it's more specific, and should be easier to mantain. I also changed the counters input to an nlp input, because there is no easy way to access the counters type of a NLPModel.

abelsiqueira commented on src/stats/stats.jl:
```diff
@@ -0,0 +1,146 @@
+                           objective :: Float64=Inf,
+                           dual_feas :: Float64=Inf,
+                           iter :: Int=-1,
+                           nlp :: AbstractNLPModel=NullNLPModel(),
```
Even better

abelsiqueira commented:
Updated and rebased. I created `AbstractExtraStats` for the specific solver stats.

dpo commented on src/stats/stats.jl:
```diff
@@ -0,0 +1,165 @@
+end
+
+function GenericExecutionStats(status :: Symbol,
+                           nlp :: AbstractNLPModel;
```
Indentation.

dpo commented on src/stats/stats.jl:
```diff
@@ -0,0 +1,165 @@
+  if !(status in keys(STATUSES))
+    s = join(keys(STATUSES, ", "))
+    error("$status is not a valid status. Use one of the following: $s")
+  end
```
This could be a one liner of the form
```
status in keys(STATUSES) || error("$status ... in $(join(keys(...)))")
```

dpo commented on src/stats/stats.jl:
```diff
@@ -0,0 +1,165 @@
+  c = Counters()
+  for counter in fieldnames(Counters)
+    setfield!(c, counter, eval(parse("$counter"))(nlp))
+  end
```
Can't we use `copy(nlp.counters)` here?

abelsiqueira replied:
Due to models like `SlackModel`, `nlp.counters` may fail. We could add a function `counters` or `copy_counters` to `NLPModels`. We actually had `counters(nlp)`, but using it to get individual counters was ugly. We could add it again, but only use it here.

abelsiqueira replied:
@dpo What do you think here? Add `counters(nlp)` on `NLPModels` again?

dpo replied:
Would it be bad to leave `counters` out of the stats type? We have methods to access the counters for each type of problem. It seems difficult to include them in stats in a generic way. Perhaps solvers could return something like `(stats, nlp.counters)` or, in the case of a SlackModel, `(stats, nlp.model.counters)`. What do you think? It's not perfect either.

abelsiqueira replied:
If `counters` is out of stats then `solve_problems` has to do the job in a similar way ([here](https://github.com/JuliaSmoothOptimizers/Optimize.jl/blob/master/src/bmark/run_solver.jl#L112)). More specifically, the return, which ends up calling the counters one by one.

> Perhaps solvers could return something like (stats, nlp.counters) or, in the case of a SlackModel, (stats, nlp.model.counters). What do you think? It's not perfect either

That would require checking inside the solver the type of `nlp`, and wouldn't work with user-defined NLPs, unless we create `counters(nlp)`.

dpo commented on src/stats/stats.jl:
```diff
@@ -0,0 +1,165 @@
+  iter :: Int
+  counters :: NLPModels.Counters
+  elapsed_time :: Float64
+  solver_specific :: AbstractExtraStats
```
On second thought, maybe `solver_specific` should just be a `Dict{Symbol,Any}` and that would simplify things?

abelsiqueira replied:
I think so too.

abelsiqueira commented:
Is there anything else to fix in this?

abelsiqueira commented:
@dpo, can we merge this?

dpo commented:
Sorry. Thank you.
